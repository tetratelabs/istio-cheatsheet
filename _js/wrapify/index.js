import matches from 'dom101/matches'
import addClass from 'dom101/add-class'
import { appendMany, nextUntil, before, findChildren, createDiv } from '../helpers/dom'

/**
 * Wraps h2 sections into h2-section.
 * Wraps h3 sections into h3-section.
 *
 * @private
 */

export default function wrapify (root) {
  // These are your H2 sections. Returns a list of .h2-section nodes.
  const sections = wrapifyH2(root)

  // For each h2 section, wrap the H3's in them
  sections.forEach(section => {
    const bodies = findChildren(section, '[data-js-h3-section-list]')
    bodies.forEach(body => { wrapifyH3(body) })
  })
}

/**
 * Wraps h2 sections into h2-section.
 * Creates and HTML structure like so:
 *
 *     .h2-section
 *       h2.
 *         (title)
 *       .body.h3-section-list.
 *         (body goes here)
 *
 * @private
 */

function wrapifyH2 (root) {
  return groupify(root, {
    tag: 'h2',
    wrapperFn: () => createDiv({ class: 'h2-section' }),
    bodyFn: () => createDiv({
      class: 'body h3-section-list',
      'data-js-h3-section-list': ''
    })
  })
}

/**
 * Wraps h3 sections into h3-section.
 * Creates and HTML structure like so:
 *
 *     .h3-section
 *       h3.
 *         (title)
 *       .body.
 *         (body goes here)
 *
 * @private
 */

function wrapifyH3 (root) {
  return groupify(root, {
    tag: 'h3',
    wrapperFn: () => createDiv({ class: 'h3-section' }),
    bodyFn: () => createDiv({ class: 'body' })
  })
}

/**
 * Groups all headings (a `tag` selector) under wrappers like `.h2-section`
 * (build by `wrapperFn()`).
 * @private
 */

export function groupify (el, { tag, wrapperFn, bodyFn }) {
  const first = el.children[0]
  let result = []

  // Handle the markup before the first h2
  if (first && !matches(first, tag)) {
    const sibs = nextUntil(first, tag)
    result.push(wrap(first, null, [ first, ...sibs ]))
  }

  // Find all h3's inside it
  const children = findChildren(el, tag)

  children.forEach(child => {
    const sibs = nextUntil(child, tag)
    result.push(wrap(child, child, sibs))
  })

  return result

  function wrap (pivot, first, sibs) {
    const wrap = wrapperFn()

    const pivotClass = pivot.className
    if (pivotClass) addClass(wrap, pivotClass)
    before(pivot, wrap)

    const body = bodyFn()
    if (pivotClass) addClass(body, pivotClass)
    appendMany(body, sibs)

    if (first) wrap.appendChild(first)
    wrap.appendChild(body)

    return wrap
  }
}
