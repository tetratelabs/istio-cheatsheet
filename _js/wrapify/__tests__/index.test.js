/* eslint-env jest */
import wrapify from '../index'
import $ from 'jquery'

it('simple usage', run(`
  <div>
    <h2>simple usage<h2>

    <h3>install</h3>
    <p>(install)</p>

    <h3>usage</h3>
    <p>(usage)</p>
  </div>
`, $div => {
  expect($div.find('.h2-section .h3-section-list .h3-section').length).toEqual(2)
}))

it('h3 with class', run(`
  <div>
    <h3 class='-hello'>install</h3>
    <p>(install)</p>
  </div>
`, $div => {
  expect($div.find('div.h3-section.-hello').length).toEqual(1)
  expect($div.find('div.h3-section-list.-hello').length).toEqual(1)
}))

it('multiple h2s', run(`
  <div>
    <h2>multiple h2<h2>

    <h3>install</h3>
    <p>(install)</p>

    <h3>usage</h3>
    <p>(usage)</p>

    <h2>getting started<h2>

    <h3>first</h3>
    <p>(first)</p>

    <h3>second</h3>
    <p>(second)</p>
  </div>
`))

function run (input, fn) {
  return function () {
    const $div = $(input)
    wrapify($div[0])
    expect($div[0]).toMatchSnapshot()
    if (fn) fn($div)
  }
}

it('h2 + pre', run(`
  <div>
    <h2>heading</h2>
    <pre class='language-markdown'>(code)</pre>
  </div>
`))
