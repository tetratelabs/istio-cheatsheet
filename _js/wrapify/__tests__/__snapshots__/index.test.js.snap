// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`h2 + pre 1`] = `
<div>
  
    
  <div
    class="h2-section"
  >
    <h2>
      heading
    </h2>
    <div
      class="body h3-section-list"
      data-js-h3-section-list=""
    >
      
    
      <div
        class="h3-section language-markdown"
      >
        <div
          class="body language-markdown"
        >
          <pre
            class="language-markdown"
          >
            (code)
          </pre>
          
  
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`h3 with class 1`] = `
<div>
  
    
  <div
    class="h2-section -hello"
  >
    <div
      class="body h3-section-list -hello"
      data-js-h3-section-list=""
    >
      <div
        class="h3-section -hello"
      >
        <h3
          class="-hello"
        >
          install
        </h3>
        <div
          class="body -hello"
        >
          
    
          <p>
            (install)
          </p>
          
  
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`multiple h2s 1`] = `
<div>
  
    
  <div
    class="h2-section"
  >
    <h2>
      multiple h2
    </h2>
    <div
      class="body h3-section-list"
      data-js-h3-section-list=""
    />
  </div>
  <div
    class="h2-section"
  >
    <h2>
      

    
    </h2>
    <div
      class="body h3-section-list"
      data-js-h3-section-list=""
    >
      <div
        class="h3-section"
      >
        <h3>
          install
        </h3>
        <div
          class="body"
        >
          
    
          <p>
            (install)
          </p>
          

    
        </div>
      </div>
      <div
        class="h3-section"
      >
        <h3>
          usage
        </h3>
        <div
          class="body"
        >
          
    
          <p>
            (usage)
          </p>
          

    
        </div>
      </div>
    </div>
  </div>
  <div
    class="h2-section"
  >
    <h2>
      getting started
    </h2>
    <div
      class="body h3-section-list"
      data-js-h3-section-list=""
    />
  </div>
  <div
    class="h2-section"
  >
    <h2>
      

    
    </h2>
    <div
      class="body h3-section-list"
      data-js-h3-section-list=""
    >
      <div
        class="h3-section"
      >
        <h3>
          first
        </h3>
        <div
          class="body"
        >
          
    
          <p>
            (first)
          </p>
          

    
        </div>
      </div>
      <div
        class="h3-section"
      >
        <h3>
          second
        </h3>
        <div
          class="body"
        >
          
    
          <p>
            (second)
          </p>
          
  
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`simple usage 1`] = `
<div>
  
    
  <div
    class="h2-section"
  >
    <h2>
      simple usage
    </h2>
    <div
      class="body h3-section-list"
      data-js-h3-section-list=""
    />
  </div>
  <div
    class="h2-section"
  >
    <h2>
      

    
    </h2>
    <div
      class="body h3-section-list"
      data-js-h3-section-list=""
    >
      <div
        class="h3-section"
      >
        <h3>
          install
        </h3>
        <div
          class="body"
        >
          
    
          <p>
            (install)
          </p>
          

    
        </div>
      </div>
      <div
        class="h3-section"
      >
        <h3>
          usage
        </h3>
        <div
          class="body"
        >
          
    
          <p>
            (usage)
          </p>
          
  
        </div>
      </div>
    </div>
  </div>
</div>
`;
